packages/backend-core/src/cache/docWritethrough.ts

```diff
@@ -0,0 +1,113 @@
+import BaseCache from "./base"
+import { getDocWritethroughClient } from "../redis/init"
+import { AnyDocument, Database, LockName, LockType } from "@budibase/types"
+import * as locks from "../redis/redlockImpl"
+
+const DEFAULT_WRITE_RATE_MS = 10000
+
+let CACHE: BaseCache | null = null
+async function getCache() {
+  if (!CACHE) {
+    const client = await getDocWritethroughClient()
+    CACHE = new BaseCache(client)
+  }
+  return CACHE
+}
+
+interface CacheItem {
+  nextWrite: number
+}
+
+export class DocWritethrough {
+  private db: Database
+  private _docId: string
+  private writeRateMs: number
+
+  private docInfoCacheKey: string
+
+  constructor(
+    db: Database,
+    docId: string,
+    writeRateMs: number = DEFAULT_WRITE_RATE_MS
+  ) {
+    this.db = db
+    this._docId = docId
+    this.writeRateMs = writeRateMs
+    this.docInfoCacheKey = `${this.docId}:info`
+  }
+
+  get docId() {
+    return this._docId
+  }
+
+  private makeNextWriteInfoItem(): CacheItem {
+    return { nextWrite: Date.now() + this.writeRateMs }
+  }
+
+  async patch(data: Record<string, any>) {
+    const cache = await getCache()
+
+    await this.storeToCache(cache, data)
+
+    const updateDb = await this.shouldUpdateDb(cache)
+
+    if (updateDb) {
+      const lockResponse = await locks.doWithLock(
+        {
+          type: LockType.TRY_ONCE,
+          name: LockName.PERSIST_WRITETHROUGH,
+          resource: this.docInfoCacheKey,
+          ttl: 15000,
+        },
+        async () => {
+          if (await this.shouldUpdateDb(cache)) {
+            await this.persistToDb(cache)
+            await cache.store(
+              this.docInfoCacheKey,
+              this.makeNextWriteInfoItem()
+            )
+          }
+        }
+      )
+
+      if (!lockResponse.executed) {
+        console.log(`Ignoring redlock conflict in write-through cache`)
+      }
+    }
+  }
+
+  private async shouldUpdateDb(cache: BaseCache) {
+    const cacheItem = await cache.withCache(this.docInfoCacheKey, null, () =>
+      this.makeNextWriteInfoItem()
+    )
+    return Date.now() >= cacheItem.nextWrite
+  }
```

Comment:
This is probably a point that applies to all our writethrough caches, each cache instance should probably have an interval, say 60 seconds, where it checks if any of the keys it has need to be written, despite no write having occurred in the last 60 seconds.

Our other writethrough caches work on the assumption that the writes will be coming at a steady rate consistently, so you can use the write as a trigger, but that doesn't work if you have bursty write rates, which SCIM logs (what this will be used for) is likely to have.

Reply:
Yup, I am checking this. This would require some background work to do these checks...